man/fi_profile.3.md

```diff
@@ -0,0 +1,222 @@
+
+## desc
+
+Reference to the description of the event. It is either defined or allocated by providers.
+
+# CALLBACK
+
+Callback functions are used to receive notifications from a provider when an event occurs. A callback function is defined by the user and called by a provider as an additional operation, normally before or after data transfer. This will increase the overhead of the operation.  Callback functions should avoid having blocking calls, or taking non-trival time. And callback functions should limit interactions with a provider to reading the variables only.
+
+The signature of the callabck function is:
+
+{% highlight c %}
+int (*callback)(struct fid_profile *prof_fid, uint32_t event,
+                void *param, size_t size, void *context)
+{% endhighlight %}
+
+The *param* is provided by the user when registering the callabck, and the *size* is data contained in the *param*.
+
+## fi_profile_open
+
+fi_profile_open is used to initiate a profiling on a target endpoint or domain specified by the fid. Upon retuning successfully, a fid_profile is returned and can be used to perform profiling operations.  If the target endpoint or domain does not support fabric profiling interface, the open call will fail with error code -FI_ENOSYS.
+
+## fi_profile_close
+
+Close the profiling operations and release related resources.
+
+## fi_profile_query_vars / fi_profile_query_events
+
+fi_profile_query_vars and fi_profile_query_events calls can be used to query available variables or events from the target endpoint or domain associated with the profile object. On input, the buffer ("varlist" or "eventlist") should be allocated, and the "count" indicates the maximum number of entries (variables or events) available in the buffer. On output, the "count" indicates the actual number of entries in the buffer. If the input "count" is smaller than the actual number of entries available, only "count" number of entries will be set in the buffer, and "-FI_ETOOSMALL" will be returned. If the buffer is NULL, upon returning, the "count" indicates the total number of entries available in the target endpoint or domain. User can allocate space and call the function again.
+
+## fi_profile_reg_callback
+
+fi_profile_reg_callback registers a callback function to the event. The event should be in the event list obtained using the fi_profile_query_events, otherwise, an error code -FI_EINVAL will be returned. 
+
+Callback function can be registered with a parameter and its size. This allows the provider to add data into the memory space pointed by the parameter when calling the callback.
+
+## fi_profile_read_u64
+
+fi_profile_read_u64 call returns an uint64_t value of the specified variable. The variable should be in the variable list obtained using fi_profile_query_vars, otherwise, an error code -FI_EINVAL will be returned. Upon returning successfully, users can use the type field in fi_profile_var_desc to cast data to an appropriate type.
```

Comment:
This should only be called for variables that are of type FI_UINT64.  The goal of having the wrapper function is that the app calls the correct wrapper based on the datatype.